packages/api/internal/dns/server.go

```diff
@@ -2,99 +2,242 @@ package dns
+func (d *DNS) cacheKey(id string) string {
+	switch {
+	case d.remote != nil:
+		return fmt.Sprintf("%s%s", cachedDnsPrefix, id)
+	case d.local != nil:
+		return fmt.Sprintf("%s.", id)
+	default:
+		return id
+	}
 }
 
-func (d *DNS) handleDNSRequest(w resolver.ResponseWriter, r *resolver.Msg) {
-	m := new(resolver.Msg)
+func (d *DNS) handleDNSRequest(ctx context.Context, w resolver.ResponseWriter, r *resolver.Msg) {
+	m := &resolver.Msg{
+		Compress: false,
+		MsgHdr: resolver.MsgHdr{
+			Authoritative: true,
+		},
+	}
+
 	m.SetReply(r)
-	m.Compress = false
-	m.Authoritative = true
 
 	for _, q := range m.Question {
 		if q.Qtype == resolver.TypeA {
-			a := &resolver.A{
+			sandboxID := strings.Split(q.Name, "-")[0]
+
+			m.Answer = append(m.Answer, &resolver.A{
 				Hdr: resolver.RR_Header{
 					Name:   q.Name,
 					Rrtype: resolver.TypeA,
 					Class:  resolver.ClassINET,
 					Ttl:    ttl,
 				},
-			}
-
-			sandboxID := strings.Split(q.Name, "-")[0]
-			ip, found := d.get(sandboxID)
-			if found {
-				a.A = net.ParseIP(ip).To4()
-			} else {
-				a.A = net.ParseIP(defaultRoutingIP).To4()
-			}
-
-			m.Answer = append(m.Answer, a)
+				A: d.Get(ctx, sandboxID),
+			})
 		}
 	}
 
-	err := w.WriteMsg(m)
-	if err != nil {
-		log.Printf("Failed to write message: %s\n", err.Error())
+	if err := w.WriteMsg(m); err != nil {
+		d.logger.Error("write DNS message", zap.Error(err))
 	}
 }
 
-func (d *DNS) Start(_ context.Context, address string, port int) error {
+var errOnStartup = errors.New("failed to start DNS server")
+
+func CheckErrOnStartup(err error) bool { return errors.Is(err, errOnStartup) }
+
+func (d *DNS) Start(ctx context.Context, address string, port int) {
+	if d.srv != nil {
+		return
+	}
+
+	// configure the underlying resolver service.
 	mux := resolver.NewServeMux()
+	mux.HandleFunc(".", func(w resolver.ResponseWriter, r *resolver.Msg) { d.handleDNSRequest(ctx, w, r) })
+	d.srv = &resolver.Server{Addr: fmt.Sprintf("%s:%d", address, port), Net: "udp", Handler: mux}
+
+	// setup error handling here: we want to catch the error from
+	// when the server starts.
+	errChan := make(chan error, 1)
+	go func() {
+		defer close(errChan)
+		if err := d.srv.ListenAndServe(); err != nil {
+			// don't do this against a context because we
+			// want this to block until everything is shut
+			// down for real, and the Close() method
+			// should work.
+			switch err.Error() {
+			case "bad network":
+				// this is the only error that can
+				// happen during startup. We have to
+				// panic here because we don't want
+				// the service to continue without any
+				// DNS service.
+				panic(errors.Join(errors.New("problem starting DNS service"), err, errOnStartup))
+			case "server already started":
+				// this only happens if you call start
+				// more than once, which shouldn't be
+				// possible.
+				errChan <- errors.Join(err, errOnStartup)
+			default:
+				// this should only happen if we
```

Comment:
I think this can happen when the port is already used

Reply:
I guess the question is "should we panic here"

if this is a "port in use error" then it's an initialization error, panicking makes sense. By not panicking we let a process run in a (potentially bad state.)

If there's a problem that happens later and the service aborts early then the process panics at an arbitrary point. This seems unlikely, but there's not a lot we can do except for... maybe a retry loop?

Reply:
I think we should panic, it will look healthy, but it won't be doing one of the most important functionalities